planner/core/rule_count_star_rewriter.go

```diff
@@ -0,0 +1,100 @@
+import (
+	"context"
+
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/expression/aggregation"
+	"github.com/pingcap/tidb/parser/mysql"
+)
+
+/**
+The countStarRewriter is used to rewrite count(*) -> count(not null column)
+
+Pattern:
+LogcialAggregation
+       |
+   DataSource
+
+Optimize:
+Table
+  <k1 bool not null, k2 int null, k3 bigint not null>
+
+Case1 there are columns exists in datasource
+Query: select count(*) from table where k3=1
+CountStarRewriterRule: pick the shortest not null column exists in datasource
+Rewritten Query: select count(k3) from table where k3=1
+
+Case2 there is no columns exists in datasource
+Query: select count(*) from table
+ColumnPruningRule: pick k1 as the shortest not null column from origin table @Function.preferNotNullColumnFromTable
+                   datasource.columns: k1
+CountStarRewriterRule: rewrite count(*) -> count(k1)
+Rewritten Query: select count(k1) from table
+
+Attention:
+Since count(*) is directly translated into count(1) during grammar parsing,
+the rewritten pattern actually matches count(1)
+*/
+
+type countStarRewriter struct {
+}
+
+func (c *countStarRewriter) optimize(ctx context.Context, p LogicalPlan, opt *logicalOptimizeOp) (LogicalPlan, error) {
+	return c.countStarRewriter(p, opt)
+}
+
+func (c *countStarRewriter) countStarRewriter(p LogicalPlan, opt *logicalOptimizeOp) (LogicalPlan, error) {
+	// match pattern agg(count(*)) -> datasource
+	if agg, ok := p.(*LogicalAggregation); ok {
+		// todo if all kinds of data source could be matched ?
+		if dataSource, ok := agg.Children()[0].(*DataSource); ok {
+			for _, aggFunc := range agg.AggFuncs {
+				if aggFunc.Name == "count" && len(aggFunc.Args) == 1 {
+					if constExpr, ok := aggFunc.Args[0].(*expression.Constant); ok {
+						if constExpr.Value.GetInt64() == 1 {
+							if len(dataSource.Columns) > 0 {
+								rewriteCount1ToCountColumn(dataSource, aggFunc)
+								continue
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+	newChildren := make([]LogicalPlan, 0, len(p.Children()))
+	for _, child := range p.Children() {
+		newChild, err := c.countStarRewriter(child, opt)
+		if err != nil {
+			return nil, err
+		}
+		newChildren = append(newChildren, newChild)
+	}
+	p.SetChildren(newChildren...)
+	return p, nil
+}
+
+// Pick the shortest and not null column from Data Source
+// If there is no not null column in Data Source, the count(1) will not be rewritten.
+func rewriteCount1ToCountColumn(dataSource *DataSource, aggFunc *aggregation.AggFuncDesc) {
+	var newAggColumn *expression.Column
+	for _, columnExistsInDataSource := range dataSource.schema.Columns {
+		if columnExistsInDataSource.IsHidden || columnExistsInDataSource.GetType().IsVarLengthType() {
+			continue
+		}
+		if mysql.HasNotNullFlag(columnExistsInDataSource.GetType().GetFlag()) {
+			if newAggColumn == nil || columnExistsInDataSource.GetType().GetFlen() < newAggColumn.GetType().GetFlen() {
+				newAggColumn = columnExistsInDataSource
+			}
```

Comment:
Is it possible that we end up choosing a wider one than rowid or pk col if all the non-hidden columns are wider fixed length column?

Reply:
make sense, should evaluate the size of pk col, especially under the clustered index is enabled.

Reply:
Changed.  The column must be narrower than rowid or pk